### PR TITLE
[RyuJIT/ARM32][LSRA] Fix restoring TYP_DOUBLE interval in ARM32

### LINKED ISSUE #11779: [RyuJIT/arm32] LinearScan::unassignPhysReg segmentation fault
```
Thread 1 "corerun" received signal SIGSEGV, Segmentation fault.
0xb0d37d8a in LinearScan::unassignPhysReg (this=0x196f44, regRec=0x1973d4, spillRefPosition=0x0) at /home/mskvortsov/git/coreclr/src/jit/lsra.cpp:6464
(gdb) bt
#0  0xb0d37d8a in LinearScan::unassignPhysReg (this=0x196f44, regRec=0x1973d4, spillRefPosition=0x0) at /home/mskvortsov/git/coreclr/src/jit/lsra.cpp:6464
#1  0xb0d3975c in LinearScan::freeRegister (this=0x196f44, physRegRecord=0x1973d4) at /home/mskvortsov/git/coreclr/src/jit/lsra.cpp:6993
#2  0xb0d39816 in LinearScan::freeRegisters (this=0x196f44, regsToFree=0) at /home/mskvortsov/git/coreclr/src/jit/lsra.cpp:7012
#3  0xb0d2fdfe in LinearScan::allocateRegisters (this=0x196f44) at /home/mskvortsov/git/coreclr/src/jit/lsra.cpp:7143
#4  0xb0d2d990 in LinearScan::doLinearScan (this=0x196f44) at /home/mskvortsov/git/coreclr/src/jit/lsra.cpp:1773
```
```
JIT/Methodical/fp/exgen/10w5d_cs_do
JIT/Methodical/fp/exgen/10w5d_cs_r
JIT/Methodical/fp/exgen/200w1d-02_cs_do
JIT/Methodical/fp/exgen/200w1d-02_cs_ro
JIT/Regression/CLR-x86-JIT/V1-M11-Beta1/b48614/b48614
JIT/Regression/CLR-x86-JIT/V1-M12-Beta2/b59782/b59782
JIT/Regression/CLR-x86-JIT/V1-M12-Beta2/b63726/b63726
JIT/Regression/CLR-x86-JIT/V1-M12-Beta2/b69225/b69225
JIT/Regression/CLR-x86-JIT/V1-M12-Beta2/b69227/b69227
JIT/Regression/JitBlue/DevDiv_283795/DevDiv_283795
```

cc @dotnet/arm32-contrib

### PR DESCRIPTION
**Pull request by hqueue**
Fix #11779 

- Fix segmentation fault by defining `nextRegRec` when `nextRegRec` is not defined yet

- Introduce `canRestorePreviousInterval()`
  This helper check whether previous interval can be restored by considering `TYP_DOUBLE` for ARM32

After this PR, segmentation fault due to uninitialized `nextRegRec` is gone with `COMPlus_AltJit=* COMPlus_AltJitAssertOnNYI=0`.
Among 10 tests reported from #11779,
- 3 tests are passed
```
JIT/Regression/CLR-x86-JIT/V1-M11-Beta1/b48614/b48614
JIT/Regression/CLR-x86-JIT/V1-M12-Beta2/b63726/b63726
JIT/Regression/CLR-x86-JIT/V1-M12-Beta2/b69225/b69225
```

- 6 tests are failed with other assertions which are reported in other issues
```
JIT/Methodical/fp/exgen/10w5d_cs_do
JIT/Methodical/fp/exgen/10w5d_cs_r
JIT/Methodical/fp/exgen/200w1d-02_cs_do
JIT/Methodical/fp/exgen/200w1d-02_cs_ro
JIT/Regression/CLR-x86-JIT/V1-M12-Beta2/b59782/b59782
JIT/Regression/CLR-x86-JIT/V1-M12-Beta2/b69227/b69227
```

- 1 test is failed due to segmentation fault due to other reason (no change by this PR)
```
JIT/Regression/JitBlue/DevDiv_283795/DevDiv_283795
```

### CODE REVIEW

**Comment by hqueue** [comment]
@CarolEidt Please take a look.

(1) Please take a look at the condition of `canRestorePreviousInterval()` for restoring previous interval for ARM32.
I've implemented a conservative approach, i.e. restore previous interval only if previous TYP_DOUBLE interval is preserved for both FP registers and both FP registers are not assigned to any interval.
(2) I didn't introduce `canRestorePreviousInterval()` for other architectures yet.
If you think this approach is reasonable, I think we can use `canRestorePreviousInterval()` for other architectures too.

**Comment by RussKeldorph** [comment]
@dotnet/jit-contrib @dotnet/arm32-contrib

File: src/jit/lsra.cpp [hunk 1]
```diff
@@ -6644,6 +6657,36 @@ bool LinearScan::isSecondHalfReg(RegRecord* regRec, Interval* interval)
...
+        (regRec->previousInterval != nullptr && regRec->previousInterval != assignedInterval &&
+         regRec->previousInterval->assignedReg == regRec && regRec->previousInterval->getNextRefPosition() != nullptr);
+
+    if (retVal && regRec->previousInterval->registerType == TYP_DOUBLE)
+    {
+        regNumber  nextRegNum = REG_NEXT(regRec->regNum);
+        RegRecord* nextRegRec = getRegisterRecord(nextRegNum);
+
+        retVal = retVal &&
+                 (nextRegRec->assignedInterval == nullptr && regRec->previousInterval == nextRegRec->previousInterval);
```
Comment (CarolEidt) [inline src/jit/lsra.cpp]: Why is the second half of this conjunction needed? If the register is unassigned (the first condition) then it is safe to reassign it to this `Interval`, even if by chance it has been assigned and unassigned to some other `Interval` (thereby potentially overwriting the `previousInterval` value). Note that the `previousInterval` handling is purely a heuristic to maintain association between an `Interval` and its preferred register.

File: src/jit/lsra.cpp [hunk 2]
```diff
@@ -6448,19 +6448,32 @@ void LinearScan::unassignPhysReg(RegRecord* regRec, RefPosition* spillRefPositio
     {
         assignedInterval->assignedReg = regRec;
     }
+#ifdef _TARGET_ARM_
+    else if (canRestorePreviousInterval(regRec, assignedInterval))
```
Comment (hqueue) [inline src/jit/lsra.cpp]: Sure. I've filed a 2nd commit.

File: src/jit/lsra.cpp (hunk 1 above)
Comment (hqueue) [inline src/jit/lsra.cpp]: I also agree and I've update the condition in a 2nd commit.

**Comment by hqueue** [comment]
@CarolEidt PTAL
- I've applied your feedback in the 2nd commit https://github.com/dotnet/coreclr/pull/11959/commits/f72202cfbb6414f1767f973797c167e77cb82923.
- I've also update logic of restoring TYP_DOUBLE for ARM32, because I found some problematic cases while updating the PR with your feedback.
 For example, when `f1` was unassigned and previousInterval of `f1` is going to be restored, than I think we should restore interval for `f0` not `f2`. Therefore I've filed 3rd commit https://github.com/dotnet/coreclr/pull/11959/commits/54f1f8014546af3e7feaca0755aa1978e288d9db.

What do you think of 3rd commit ?

**Pull request merged**